src/VisualStudio/Core/Def/Implementation/InheritanceMargin/InheritanceGlyphManager.cs

```diff
@@ -0,0 +1,186 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Windows.Controls;
+using System.Windows.Media;
+using Microsoft.CodeAnalysis.Editor.Host;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.VisualStudio.LanguageServices.Implementation.InheritanceMargin.MarginGlyph;
+using Microsoft.VisualStudio.PlatformUI;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Classification;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Formatting;
+using Microsoft.VisualStudio.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.InheritanceMargin
+{
+    /// <summary>
+    /// Manager controls all the glyph of Inheritance Margin
+    /// </summary>
+    internal class InheritanceGlyphManager
+    {
+        private readonly IWpfTextView _textView;
+        private readonly IThreadingContext _threadingContext;
+        private readonly IStreamingFindUsagesPresenter _streamingFindUsagesPresenter;
+        private readonly ClassificationTypeMap _classificationTypeMap;
+        private readonly IClassificationFormatMap _classificationFormatMap;
+        private readonly IUIThreadOperationExecutor _operationExecutor;
+        private readonly IEditorFormatMap _editorFormatMap;
+        private Dictionary<InheritanceMarginGlyph, SnapshotSpan> _glyphToTaggedSpan;
```

Comment:
not readonly?

Reply:
The value is later changed in SetSnapshotAndUpdate() method so it's not readonly

Reply:
hrmm... can it be readonly, and you clear it and add to it?  that would be nicer instead of making fresh instances.  if you do intend to make fresh instances, can you make this an IMmutableDictionary instead? i don't like it when both the field and he data are both mutable.  harder to reason about.

also, if this can be mutated (either the data, or the field), please doc the threading constraints around that.